doc/modules/mod_mam.md

```diff
@@ -145,6 +145,61 @@ You can change the default settings using extra parameters:
 }.
 ```
 
+### Custom backend
+
+You can implement a custom backend for mod_mam. In this case the module needs to implement the `gen_mod` behavior and make it have its own toplevel place within the modules section of your configuration file, eg:
+
+``` erlang
+{modules, [
+  ...
+  {mod_mam_custom_arch, []},
+  {mod_mam, [...]},
+  ...
+]}
+```
+
+Typically your module would add handlers for hooks triggered by mod_mam. See other backend implementations to get an idea.
+
+Eg.:
+
+``` erlang
+start(Host, _Opts) ->
+  ...
+  ejabberd_hooks:add(mam_lookup_messages, Host, ?MODULE, lookup_messages, 50),
+  ...
+  ok.
+
+...
+
+-spec lookup_messages(any(), ejabberd:server(), map()) -> {ok, mod_mam:lookup_result()} |
+                                                          {error, 'policy-violation'} |
+                                                          {error, any()}.
+lookup_messages({error, _Reason} = Result, _Host, _Params) -> Result;
+lookup_messages(_Result, Host, #{owner_jid        := OwnerJid,
+                                 rsm              := RSM,
+                                 start_ts         := undefined,
+                                 end_ts           := undefined,
+                                 with_jid         := undefined,
+                                 search_text      := undefined,
+                                 page_size        := PageSize,
+                                 limit_passed     := _LimitPassed,    % ?
```

Comment:
there is logic in MAM 0.3. (the version, that does not use forms for lookups).

- if an user provides a limit of messages in his request, than we are perfectly fine
- if an user does not provide the limit, and there are a lot of messages matching, the server CAN return `policy-violation` error (see in spec!).

So, yep, limit_passed flag just makes backends logic harder. I would rather make the logic in mod_mam/mod_mam_muc.

max_result_limit is `min(Limit_specified_by_user, Limit_specified_in_mongoose_config)`.